File: 010_sparse_matrices/solution/triangles_matrix.m
function A = triangles_matrix(n)
%TRIANGLES_MATRIX Construct a n x n sparse matrix with a triangular pattern
%of ones, as by the following pattern:
% 0 1 1 1 1 1 1 1 1 1 1
% 1 0 1 0 0 0 0 0 0 0 1
% 1 1 0 1 0 0 0 0 0 0 1
% 1 0 1 0 1 0 0 0 0 0 1
% 1 0 0 1 0 1 0 0 0 0 1
% 1 0 0 0 1 0 1 0 0 0 1
% 1 0 0 0 0 1 0 1 0 0 1
% 1 0 0 0 0 0 1 0 1 0 1
% 1 0 0 0 0 0 0 1 0 1 1
% 1 0 0 0 0 0 0 0 1 0 1
% 1 1 1 1 1 1 1 1 1 1 0
%
% t = four_corners(m,n);
%
% Inputs:
%  n  the dimension of the matrix requested
% Outputs:
%  A  the sparse matrix with the requested triangle pattern
%

i = [(2:n)'; (1:(n-1))'; ones(n-2,1); n*ones(n-2,1); ...
    (3:(n-1))'; (2:(n-2))'];

j = [ones(n-1,1); n*ones(n-1,1); (2:(n-1))'; (2:(n-1))'; ...
    (2:(n-2))'; (3:(n-1))']; 

v = ones(size(i));

A = sparse(i, j, v, n, n);

end

File: 010_sparse_matrices/solution/triangles_matrix.py
import numpy as np, scipy as sp

def triangles_matrix(n):
    """This function constructs an n x n sparse matrix with a triangular pattern
    of ones, as by the following pattern:
    0 1 1 1 1 1 1 1 1 1 1
    1 0 1 0 0 0 0 0 0 0 1
    1 1 0 1 0 0 0 0 0 0 1
    1 0 1 0 1 0 0 0 0 0 1
    1 0 0 1 0 1 0 0 0 0 1
    1 0 0 0 1 0 1 0 0 0 1
    1 0 0 0 0 1 0 1 0 0 1
    1 0 0 0 0 0 1 0 1 0 1
    1 0 0 0 0 0 0 1 0 1 1
    1 0 0 0 0 0 0 0 1 0 1
    1 1 1 1 1 1 1 1 1 1 0
    """

    i = np.concatenate((np.arange(1,n), np.arange(0,n-1),
                        np.zeros(n-2), (n-1)*np.ones(n-2),
                        np.arange(2,n-1), np.arange(1,n-2)))
    j = np.concatenate((np.zeros(n-1), (n-1)*np.ones(n-1),
                        np.arange(1,n-1), np.arange(1,n-1),
                        np.arange(1,n-2), np.arange(2,n-1)))
    k = np.ones(i.shape[0])
    A = sp.sparse.csr_matrix((k,(i,j)), shape=(n,n))

    return A


